File: fc-python/FCweather.py
import openai
import json
from openai import OpenAI
import requests
import sys
from config import OPENAI_API_KEY, BASE_URL, MODEL_NAME


latitude=None
longtitude=None

    # Read API key
    # with open("api.txt", "r") as f:
    #     api_key = f.read().strip()
    # print("API key loaded successfully")
    
        # Initialize client with base URL
client = OpenAI(
    api_key="sk-wosxiisuzqcpwbnmaobpgflmgxzpumvxsuvusoduscvhcdoc",
    base_url="https://api.siliconflow.cn/v1"
)
print("OpenAI client initialized")

def get_weather( latitude: float, longtitude: float):
    api_key = "ff3f2807abcb8a97b30115d180fdde54"  # Replace with your actual OpenWeatherMap API key
    base_url = "https://api.openweathermap.org/data/2.5/weather"
    params = {
        'lat': latitude,
        'lon': longtitude,
        'appid': api_key
    }

    response = requests.get(base_url, params=params)

    if response.status_code == 200:
        data = response.json()
        weather = data['weather'][0]['description']
        temperature = data['main']['temp']-273.15
        feels_like = data['main']['feels_like']-273.15
        humidity = data['main']['humidity']
        wind_speed = data['wind']['speed']
        return f"The weather in coordinates {latitude} and {longtitude} is {weather} with a temperature of {temperature}celsius (feels like {feels_like}celsius), humidity {humidity}%, and wind speed {wind_speed} m/s"
    else:
        return f"Sorry, I could not retrieve the weather for coordinates {latitude} and {longtitude} at the moment"


def get_coordinates(city_name: str):
    print(f"getting coordinates for city:{city_name}")
    api_key="ff3f2807abcb8a97b30115d180fdde54"
    base_url="https://api.openweathermap.org/geo/1.0/direct"
    params={
        "q":city_name,
        "appid":api_key
    }
    
    response=requests.get(base_url,params=params)
    if response.status_code==200:
        data=response.json()
        if data:
            latitude=data[0]['lat']
            longtitude=data[0]['lon']
            latitude=float(latitude)
            longtitude=float(longtitude)
            return latitude,longtitude
        else:
            return None,None
        
    else:
        print(f"Error:{response.status_code}")
        return None,None
    


tools = [
    {
        'type': 'function',
        'function': {
            'name': 'get_weather',
            'description': 'Get the weather for a given coordinates',
            'parameters': {
                'type': 'object',
                'properties': {
                    'latitude': {
                        'type': 'number',
                        'description': 'latitude of the location'
                    },
                    'longtitude': {
                        'type': 'number',
                        'description': 'longitude of the location'
                    }
                },
                'required': ['latitude', 'longtitude'],
                'additionalProperties': False
            }
        }
    },
    {
        'type': 'function',
        'function': {
            'name': 'get_coordinates',
            'description': 'Get the coordinates of a given city name, eg: "New York"',
            'parameters': {
                'type': 'object',
                'properties': {
                    'city_name': {
                        'type': 'string',
                        'description': 'the name of the city provided by the user to fetch corresponding coordinates',
                    }
                },
                'required': ['city_name']
            }
        }
    }
]



def function_call_playground_coords(prompt: str):
    print("starting function call playground for coordinates")
    global latitude, longtitude
    messages = [
        {'role': 'system', 'content': prompt}
    ]
    response = client.chat.completions.create(
        model="THUDM/GLM-4-9B-0414",
        messages=messages,
        tools=tools,
        temperature=0.01,
        stream=False,
        top_p=0.95,
    )

    func_call = response.choices[0].message.tool_calls[0]
    func1_name = func_call.function.name
    func1_args = json.loads(func_call.function.arguments)  # safer parsing
    func1_out = eval(f'{func1_name}(**{func1_args})')

    # Only set global vars if successful
    if func1_out and func1_out[0] is not None and func1_out[1] is not None:
        latitude, longtitude = func1_out
    else:
        print("Failed to retrieve coordinates.")
        latitude, longtitude = None, None

    messages.append(response.choices[0].message)
    messages.append({
        'role': 'tool',
        'content': f'Coordinates: {latitude}, {longtitude}',
        'tool_call_id': func_call.id
    })

    return f"Coordinates obtained: {latitude}, {longtitude}" if latitude is not None else "Failed to get coordinates."





def function_call_playground(prompt: str):
    print("starting function call playground for weather")
    global latitude, longtitude
    messages = [{"role": "system", "content": prompt}]

    response = client.chat.completions.create(
        model="THUDM/GLM-4-9B-0414",
        messages=messages,
        temperature=0.01,
        stream=False,
        top_p=0.95,
        tools=tools
    )

    tool_calls = response.choices[0].message.tool_calls

    if tool_calls is None:
        print("⚠️ No tool call was made by the model.")
        return "No tool call triggered."

    func_call = tool_calls[0]
    func_name = func_call.function.name
    func_args = json.loads(func_call.function.arguments)
    func_out = eval(f'{func_name}(**{func_args})')

    if func_name == 'get_coordinates':
        if func_out and func_out[0] is not None:
            latitude, longtitude = func_out
            result = f"Coordinates obtained: {latitude}, {longtitude}"
        else:
            result = "Failed to retrieve coordinates."
    elif func_name == 'get_weather':
        result = func_out  # Just a string message
    else:
        result = "Unknown function returned."

    messages.append(response.choices[0].message)
    messages.append({
        'role': 'tool',
        'content': str(func_out),
        'tool_call_id': func_call.id
    })

    final_response = client.chat.completions.create(
        model="THUDM/GLM-4-9B-0414",
        messages=messages,
        temperature=0.01,
        stream=False,
        top_p=0.95,
        tools=tools
    )

    return final_response.choices[0].message.content or result


prompt1="what are the coordinates of Zurich, Switzerland?"

import time
start_time=time.time()

# First get coordinates
print(function_call_playground_coords(prompt1))
print(f"coordinates:{latitude},{longtitude}")
# Then get weather using the stored coordinates
if latitude is not None and longtitude is not None:
    prompt2 = f"What is the weather like in coordinates {latitude}, {longtitude} today?"
    print(function_call_playground(prompt2))
else:
    print("No weather request made due to missing coordinates.")


end_time=time.time()
print(f"Execution time:{end_time-start_time} seconds")

File: fc-nodejs/FCweather.js
import { OpenAI } from "openai";
import { config } from "dotenv";
import {BASE_URL, OPENAI_API_KEY, MODEL_NAME, OPENWEATHERMAP_API_KEY} from "./config.js";

config();
let longtitude = 0;
let latitude = 0;
// console.log("Environment variables:");
// console.log("OPENAI_API_KEY:", process.env.OPENAI_API_KEY);
// console.log("MODEL_NAME:", process.env.MODEL_NAME);
// console.log("BASE_URL:", "https://api.siliconflow.cn/v1");

const openai = new OpenAI({
    apiKey: OPENAI_API_KEY,
    baseURL: BASE_URL,
});

const getCoordinates = async(city)=>{
    console.log("get coordinates started for city:",city);
    const apiKey = OPENWEATHERMAP_API_KEY;
    const base_url = "https://api.openweathermap.org/geo/1.0/direct";
    const params = {
        q: city,
        appid: apiKey
    };
    const response = await fetch(`${base_url}?${new URLSearchParams(params)}`);
    if(response.status === 200){
        const data = await response.json();
        if(data.length > 0){
            const {name, lat, lon} = data[0];
            console.log("coordinates fetched for city:",name, lat, lon);
            return {name, lat, lon};
        }
    }
    console.error("failed to fetch coordinates with status code:", response.status);
    return null;
}

const getWeather = async(latitude, longtitude) => {
    console.log("get weather started for coordinates:",latitude, longtitude);
    const apiKey = OPENWEATHERMAP_API_KEY;
    const url = "https://api.openweathermap.org/data/2.5/weather";
    const params = new URLSearchParams({
        lat: latitude,
        lon: longtitude,
        appid: apiKey,
    });

    const response = await fetch(`${url}?${params}`);
    if(response.status === 200){
        const data = await response.json();
        const weather = data.weather[0].description;
        return `the weather at ${latitude} and ${longtitude} is ${weather}`;
    } else {
        return `failed to fetch weather at coordinates ${latitude} and ${longtitude}`;
    }
}

const tools = [{
    type: "function",
    function: {
        name: "getWeather",
        description: "get the weather for a specific latitude and longtitude",
        parameters: {
            type: "object",
            properties: {
                latitude: {
                    type: "number",
                    description: "the latitude of the location",
                },
                longtitude: {
                    type: "number",
                    description: "the longtitude of the location",
                }
            },
            required: [
                "latitude",
                "longtitude",
            ],
        },
    }
},
{
    type:"function",
    function:{
        name:"getCoordinates",
        description:"get the coordinates for a specific city",
        parameters:{
            type:"object",
            properties:{
                city:{
                    type:"string",
                    description:"the name of the city to get coordinates for",
                }
            },
            required:[
                "city",
            ]
        }
    },
}
]
export const function_call_playground_coordinates = async(prompt)=>{
    console.log("function call playground for coordinates starting...");
    const messages=[
        {
            role:"system",
            content:prompt,
        }
    ];
    const response = await openai.chat.completions.create({
        model:MODEL_NAME,
        messages:messages,
        temperature:0.01,
        stream:false,
        top_p:0.95,
        tools:tools,
    })

    const func1_name = response.choices[0].message.tool_calls[0].function.name;
    const func1_args = JSON.parse(response.choices[0].message.tool_calls[0].function.arguments);
    const func1_out = await getCoordinates(func1_args.city);

    if(func1_out && func1_out.lat && func1_out.lon){
        console.log("coordinates fetched for city", func1_out.lat, func1_out.lon);
        latitude = func1_out.lat;
        longtitude = func1_out.lon;
        return {latitude: func1_out.lat, longitude: func1_out.lon};
    } else {
        console.error("failed to fetch coordinates for city:", func1_args.city);
        return null;
    }
}

export const function_call_playground_weather = async(prompt) => {
    console.log("function call playground for weather starting...");
    const messages = [
        {
            role: "system",
            content: prompt,
        },
    ];

    const response = await openai.chat.completions.create({
        model: MODEL_NAME,
        messages: messages,
        temperature: 0.01,
        stream: false,
        top_p: 0.95,
        tools: tools,
    })
    if (!response.choices[0].message.tool_calls){
        console.error("no tool calls found in the response");
        return null;
    }
    const func1_name = response.choices[0].message.tool_calls[0].function.name;
    const func1_args = JSON.parse(response.choices[0].message.tool_calls[0].function.arguments);
    const func1_out = await getWeather(latitude, longtitude);

    messages.push(response.choices[0].message);
    messages.push({
        role: "tool",
        content: `${func1_out}`,
        tool_call_id: response.choices[0].message.tool_calls[0].id,
    })

    const response2 = await openai.chat.completions.create({
        model: MODEL_NAME,
        messages: messages,
        temperature: 0.01,
        stream: false,
        top_p: 0.95,
        tools: tools,
    })

    return response2.choices[0].message.content;
}

const prompt1 = "what are the coordinates of beijing, china?";

// Execute the function
const start_time = Date.now();

// First get coordinates
function_call_playground_coordinates(prompt1)
    .then(coords => {
        if (!coords) {
            throw new Error("Failed to get coordinates");
        }
        console.log("Got coordinates:", coords);
        // Create prompt2 here after we have the coordinates
        const prompt2 = `What is the weather like at latitude ${coords.latitude} and longitude ${coords.longitude}?`;
        return function_call_playground_weather(prompt2);
    })
    .then(result => {
        const end_time = Date.now();
        console.log(`Time taken for function call playground: ${(end_time - start_time) / 1000} seconds`);
        console.log(result);
    })
    .catch(error => {
        console.error("Error:", error);
    });
